### PR TITLE
Feature/#142 update topic api

### PR DESCRIPTION
**Pull request by kgingstad**
- only allows topics that a user has never seen before to be submitted (not previously recommended or manually added)
- extended user_feedback/topics endpoint to also return topic states
- documented changes

### CODE REVIEW

File: arxivdigest/api/README.md [hunk 1]
```diff
@@ -475,7 +487,7 @@ Example:
 
 `POST /recommendations/topics`
 
-Insert recommendations for topics to users, with a score describing how well it matches the users interests. Sending the same recommendation multiple times will update the score to the last received value. This allows reordering of already submitted recommendations, but assumes comparable scores across submissions. See the [recommendation submission guide](/../../#howto-for-experimental-recommender-systems) for more information on how to submit recommendations.   
+Insert recommendations for topics to users, with a score describing how well it matches the users interests. Will only accept topics that have not been recommended at an earlier stat or is added manually by a user. See the [recommendation submission guide](/../../#howto-for-experimental-recommender-systems) for more information on how to submit recommendations.
```
Comment (oyvi96) [inline arxivdigest/api/README.md:490]: Line is a bit hard to understand.

File: arxivdigest/api/README.md [hunk 2]
```diff
@@ -234,6 +234,15 @@ Fields returned for each user:
   - `feedback`: is the feedback stored in a dictionary of {feedback_type: datetime}
     -  "seen":   datetime of when article was seen or null if not seen
     -  "clicked":     datetime of when topic was clicked or null if not clicked
+    -  "state": USER_ADDED, USER_REJECTED, EXPIRED, REFRESHED, SYSTEM_RECOMMENDED_ACCEPTED, SYSTEM_RECOMMENDED_REJECTED
+
+Explanations of different states:
+  - USER_ADDED: The user added the topic themselves by writing it.
+  - USER_REJECTED: The user removed a topic from their profile that was previously USER_ADDED.
+  - EXPIRED: The topic was recommended, but the user did not interact with it within 24 hours.
```
Comment (kbalog) [inline arxivdigest/api/README.md:242]: "within 24 hours of seeing this recommendation."

File: arxivdigest/api/database.py [hunk 3]
```diff
@@ -252,6 +252,29 @@ def get_topic_recommendations(ids):
 
         return users
 
+def get_user_topics(ids):
```
Comment (oyvi96) [inline arxivdigest/api/database.py:255]: should be merged with get_user_feedback_topics such that the same topics that are blocked are sent to the systems, such that they can know upfront which topics are blocked.

This function should probably not exist.

File: arxivdigest/api/README.md [hunk 4]
```diff
@@ -234,6 +234,15 @@ Fields returned for each user:
...
+  - REFRESHED: The topic was recommended, but the user refreshed the topic suggestion list without interacting with it.
+  - SYSTEM_RECOMMENDED_ACCEPTED: The topic was recommended and the user accepted it from the list or manually added it after rejecting it.
```
Comment (kbalog) [inline arxivdigest/api/README.md:244]: If manually added after rejecting a system recommendation, then the state should be USER_ADDED.

File: arxivdigest/api/validator.py [hunk 5]
```diff
@@ -90,6 +91,28 @@ def topic_recommendation(json):
             return err
     return None
 
+def duplicate_topic_suggestion(json):
```
Comment (oyvi96) [inline arxivdigest/api/validator.py:94]: should use same logic for blocking as get_user_feedback_topics such that we ensure consistency

File: arxivdigest/api/README.md (hunk 1 above)
Comment (kbalog) [inline arxivdigest/api/README.md:490]: ... have not been recommended _before_ or is ...

File: arxivdigest/api/database.py [hunk 6]
```diff
@@ -300,22 +323,28 @@ def get_user_feedback_topics(ids):
...
         ], ...
     }
     """
     with closing(getDb().cursor(dictionary=True)) as cur:
-        sql = """SELECT user_id, topic, DATE(datestamp) as date, seen, clicked
-                 FROM topic_recommendations tr NATURAL JOIN topics t
-                 WHERE user_id IN ({}) AND interleaving_order IS NOT 
-                 null ORDER BY  interleaving_order 
-                 DESC""".format(','.join(['%s'] * len(ids)))
+        sql = '''SELECT tr.user_id, t.topic, DATE(tr.datestamp) as date,
```
Comment (oyvi96) [inline arxivdigest/api/database.py:333]: This function should also return data/feedback on user added topics, such that systems does not recommend them again.

File: arxivdigest/api/database.py [hunk 7]
```diff
@@ -252,6 +252,29 @@ def get_topic_recommendations(ids):
...
+        sql = '''SELECT temp.user_id, t.topic FROM                
+            (SELECT tr.user_id, tr.topic_id FROM  topic_recommendations tr
+            LEFT JOIN user_topics ut ON ut.topic_id = tr.topic_id and 
+            tr.user_id = ut.user_id WHERE tr.user_id in (%s)
+            UNION
+            SELECT ut.user_id, ut.topic_id FROM  topic_recommendations tr
+            RIGHT JOIN user_topics ut ON ut.topic_id = tr.topic_id and 
+            tr.user_id = ut.user_id WHERE tr.user_id in (%s)) temp
+            LEFT JOIN topics t 
+            on t.topic_id=temp.topic_id''' % (placeholders, placeholders)
```
Comment (oyvi96) [inline arxivdigest/api/database.py:270]: Should only select topics where interleaving order is not null or is user_added or user_rejected.

Several systems can recommend the same topics until the user has interacted with it.

File: arxivdigest/api/README.md [hunk 8]
```diff
@@ -475,7 +487,7 @@ Example:
 
 `POST /recommendations/topics`
 
-Insert recommendations for topics to users, with a score describing how well it matches the users interests. Sending the same recommendation multiple times will update the score to the last received value. This allows reordering of already submitted recommendations, but assumes comparable scores across submissions. See the [recommendation submission guide](/../../#howto-for-experimental-recommender-systems) for more information on how to submit recommendations.
```
[the remark this answers is not in the document]
Comment (kgingstad) [reply arxivdigest/api/README.md:478]: I just removed this bit:
`Sending the same recommendation multiple times will update the score to the last received value. This allows reordering of already submitted recommendations, but assumes comparable scores across submissions.`
Replaced it with this:
`Will only accept topics that have not been recommended at an earlier stat or is added manually by a user.`

[the remark this answers is not in the document]
Comment (oyvi96) [reply arxivdigest/api/README.md:478]: Yeah but I believe that the previous line should still be there.

[the remark this answers is not in the document]
Comment (kgingstad) [reply arxivdigest/api/README.md:478]: but it does not allow the same topics to be recommended twice so the score cant be updated?

File: arxivdigest/api/database.py (hunk 3 above)
Comment (kgingstad) [reply arxivdigest/api/database.py:255]: problem is the user feedback only returns topics that have been interleaved, where this function also needs to return the topics that have been recommended but not yet been interleaved

File: arxivdigest/api/validator.py [hunk 9]
```diff
@@ -90,6 +91,30 @@ def topic_recommendation(json):
...
+    for that user. Returns error message if there are topics 
+    that have been suggested previously and status code."""
+    user_ids = [user_id for user_id in json]
+    prev_topics = db.get_user_feedback_topics(user_ids)
+    error_msg = ('Some of the recommended topics have '
+                'already been recommended for users:')
+    error = False
+    for user_id in user_ids:
+        new_topics = set([topic['topic'] for topic in json[user_id]])
+        print(new_topics, '\n')
```
Comment (oyvi96) [inline arxivdigest/api/validator.py:105]: debug print statements in final commit

File: arxivdigest/api/README.md [hunk 10]
```diff
@@ -219,21 +219,29 @@ Fields are returned in a JSON in the format, the topics are sorted descending by
     {
       "user_feedback": {
         user_id: {
-           date: [
-            {topic: feedback},
-            {topic: feedback},
-           ]
+          {topic: feedback},
```
Comment (oyvi96) [inline arxivdigest/api/README.md:222]: not valid json, probably should be something like
user_id: {
              topic: feedback_dict,
              topic: feedback_dict,
             }

File: arxivdigest/api/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (oyvi96) [reply arxivdigest/api/README.md:478]: Still removed

File: arxivdigest/api/database.py [hunk 11]
```diff
@@ -295,27 +294,42 @@ def get_user_feedback_articles(ids):
...
+        }, ...
     }
     """
     with closing(getDb().cursor(dictionary=True)) as cur:
-        sql = """SELECT user_id, topic, DATE(datestamp) as date, seen, clicked
-                 FROM topic_recommendations tr NATURAL JOIN topics t
-                 WHERE user_id IN ({}) AND interleaving_order IS NOT 
-                 null ORDER BY  interleaving_order 
-                 DESC""".format(','.join(['%s'] * len(ids)))
+        tr_sql = '''SELECT tr.user_id, t.topic, tr.datestamp
```
Comment (oyvi96) [inline arxivdigest/api/database.py:309]: should include interleaving order in result, as its generally not okay to trust the sort order of a dict.

File: arxivdigest/api/README.md [hunk 12]
```diff
@@ -259,33 +259,33 @@ Example request:
...
             }
-          ]
+          },
+          {
+            "transportation planning": {
+              "clicked": null,
+              "seen": "2020-03-17 17:13:53",
+              "state": "REFRESHED",
+              "recommendation_time": "2020-03-15 11:16:53"
+              "interleavnig_order": 4
```
Comment (kbalog) [inline arxivdigest/api/README.md:278]: typo "interleaving"

**Pull request merged**